packages/ts/file-router/src/vite-plugin.ts

```diff
@@ -83,40 +87,80 @@ export default function vitePluginFileSystemRouter({
         code: new URL('file-routes.ts', _generatedDir),
         layouts: new URL('layouts.json', _generatedDir),
       };
+      _generateRuntimeFiles = async (): Promise<boolean> => {
+        try {
+          return await generateRuntimeFiles(_viewsDir, runtimeUrls, extensions, _logger, debug);
+        } catch (e: unknown) {
+          _logger.error(String(e));
+          return true;
+        }
+      };
     },
     async buildStart() {
-      try {
-        await generateRuntimeFiles(_viewsDir, runtimeUrls, extensions, _logger, debug);
-      } catch (e: unknown) {
-        _logger.error(String(e));
-      }
+      await _generateRuntimeFiles();
     },
-    configureServer(server) {
-      const dir = fileURLToPath(_viewsDir);
-
-      const changeListener = (file: string): void => {
-        if (!file.startsWith(dir)) {
-          if (file === fileURLToPath(runtimeUrls.json)) {
-            server.hot.send({ type: 'custom', event: 'fs-route-update' });
-          } else if (file !== fileURLToPath(runtimeUrls.layouts)) {
-            // outside views folder, only changes to layouts file should trigger files generation
-            return;
-          }
+    async hotUpdate({ file, modules }: HotUpdateOptions): Promise<void | EnvironmentModuleNode[]> {
+      const fileUrlString = String(pathToFileURL(file));
+
+      if (fileUrlString === String(runtimeUrls.json)) {
+        // Reload file routes JSON with a custom HMR event.
+        this.environment.hot.send({ type: 'custom', event: 'fs-route-update' });
+        return [];
+      }
+
+      if (fileUrlString === String(runtimeUrls.code)) {
+        // Skip HMR for file routes from Vite builtin file change listener,
+        // as we have our own HMR handling below.
+        return [];
+      }
+
+      if (!file.startsWith(_viewsDirUsingSlashes)) {
+        // Outside views folder, only changes to layouts file should trigger
+        // files generation.
+        if (fileUrlString !== String(runtimeUrls.layouts)) {
+          return;
         }
+      }
 
-        generateRuntimeFiles(_viewsDir, runtimeUrls, extensions, _logger, debug).catch((e: unknown) =>
-          _logger.error(String(e)),
+      // Check and update file routes if needed.
+      if (await _generateRuntimeFiles()) {
+        // The "file-routes.ts" file was changed at this point, so it should
+        // be considered within the HMR update that caused the routes update.
+
+        const mg = this.environment.moduleGraph;
+        const fileRoutesModules = mg.getModulesByFile(fileURLToPath(runtimeUrls.code).replaceAll('\\', '/'))!;
```

Comment:
```suggestion
        const fileRoutesModules = mg.getModulesByFile(fileURLToPath(runtimeUrls.code, { windows: false }))!;
```

Reply:
`windows` option is Node 22+. Since Vaadin still supports Node 20, let us not use it just yet.